// djed/marketing/models/Stories.js
'use strict';

var _ = require( 'lodash' );
var Collection = require( 'djed/interface/model/Collection' );

module.exports = Collection.extend( {

  initialize: function ( models, options ) {
    this.active = _.findIndex( models, function ( model ) {
      return options.active === model.id;
    } );
  },

  goLeft: function () {
    this.active = this.active === 0 ? this.length - 1 : this.active - 1;
    this.update();
  },

  goRight: function () {
    this.active = this.active === this.length - 1 ? 0 : this.active + 1;
    this.update();
  },

  switch: function ( newActive ) {
    newActive = parseInt( newActive );
    if ( newActive !== this.active ) {
      this.active = newActive;
      this.update();
    }
  },

  update: function () {
    this.trigger( 'switch', this.at( this.active ).id );
  },

  getActiveContent: function () {
    return this.at( this.active );
  },

  getSurroundingContent: function () {
    var active = this.at( this.active );
    if ( this.active === 0 ) {
      return [ active !== this.last() && this.last(), active, this.at( this.active + 1 ) ];
    } else if ( this.active === this.length - 1 ) {
      return [ this.at( this.active - 1 ), active, active !== this.at( 0 ) && this.at( 0 ) ];
    } else {
      return this.slice( this.active - 1, this.active + 2 );
    }
  }

} );
