// djed/marketing/components/ProductShot.js
'use strict';

var View = require( 'djed/interface/view/View' );

module.exports = View.extend( {

  initialize: function ( options ) {
    this.slideTime = options.slideTime;
    this.listenTo( this.model, 'switch', this.switch );
    this.inview = new window.Waypoint.Inview( {
      element: this.$el,
      enter: this.transitionIn.bind( this )
    } );

  },

  transitionIn: function () {
    var elements = this.getElements();
    if ( !this.hasTransitioned ) {
      this.hasTransitioned = true;
      // TODO animate on entry
    }
  },

  getElements: function () {
    return {
      browser: this.$( '.product-browser' ),
      phone: this.$( '.product-iphone' ),
      headline: this.$( 'h2' ),
      subhead: this.$( 'p.subhead' )
    };
  },

  switch: function () {
    var data = this.model.getActiveContent().get( 'product-shot' );
    var elements = this.getElements();
    setTimeout( function () {
      elements.browser.find( 'img' ).attr( 'src', data[ 'browser-imagepath' ] );
      elements.phone.attr( 'src', data[ 'phone-imagepath' ] );
      elements.headline.text( data.headline );
      elements.subhead.text( data.subhead );
    }, this.slideTime / 2 );
  },

  remove: function () {
    this.inview.destroy();
    View.prototype.remove.apply( this );
  }

} );
